Cleanfigure

theRealSuperMario opened the pull request:
This PR addresses #258 by introducing a `cleanfigure` command. Essentially, I translated the awesome [matlab2tikz](https://github.com/matlab2tikz/matlab2tikz) code for cleaning up figures into python for matplotlib.

The code still has some unimplemented code blocks which will raise an error. However, I heavily tested my code and could never reach those blocks. So I decided to leave it like that for now and make the PR.

In short, it works for `plt.plot` and `plt.plot3d` commands, which are probably the most common scenarios.


It would be great to get some feedback on the code and what is still missing to merge the PR.

theRealSuperMario commented:
I'm a little confused on how to get the style tests to work. It looks like black and flake8 have conflicting conventions on whitespaces after commas, like `[1,]` vs `[1, ]`.

nschloe commented:
All of the lint errors are `blank line contains whitespace`, aren't they? Just remove all trailing whitespace and you should be fine.

theRealSuperMario commented:
> All of the lint errors are `blank line contains whitespace`, aren't they? Just remove all trailing whitespace and you should be fine.

I fixed the `blank line contains whitespace` errors, but fixing `E231 missing whitespace after ','` is not possible because black eliminates the space again.

Example:
```bash

# line after black formatting
np.allclose(mask * 1, np.array([1,] + [0,] * 10 + [1,]))

# what flake8 would like to have
np.allclose(mask * 1, np.array([1, ] + [0, ] * 10 + [1, ]))
```

I have noticed that there are already some ignored rules in `.flake8`, maybe just add `E231`?

theRealSuperMario commented:
> All of the lint errors are `blank line contains whitespace`, aren't they? Just remove all trailing whitespace and you should be fine.

Got it. Thanks. Looks like I need to write some more tests to get the coverage right.

nschloe commented:
This is looking really good.

For the tests, I see that you're testing the internal (underscored) functions. This is overkill. I'd rather see the user functions tested here, i.e., just `cleanfigure`.

theRealSuperMario commented:
> For the tests, I see that you're testing the internal (underscored) functions. This is overkill. I'd rather see the user functions tested here, i.e., just `cleanfigure`.

I would argue against that. I wrote the code by stepping through the `matlab2tikz` code and making sure every subroutine of the python code gives exactly the same output as the `matlab2tikz` code. I found it very useful to have this fine-grained control, also when adding more and more features.

Since not all functions are implemented yet, keeping all the tests makes sure that we don't screw up some subroutine in the future.


I just added some more tests to increase the coverage.

nschloe commented:
In general one should not test implementation details. I don't want to be too strict about this now, let's keep it in if you want.

> I was just thinking if a Readme update would be good to point to the new cleanfigure functionality.

Sure. A small remark is enough here I think.

theRealSuperMario commented:
> In general one should not test implementation details. I don't want to be too strict about this now, let's keep it in if you want.

Hmm, yeah I see your point. I suggest to remove them once all the functionality has been implemented. For instance, cleaning scatter plots is not possible yet.


> > I was just thinking if a Readme update would be good to point to the new cleanfigure functionality.
> 
> Sure. A small remark is enough here I think.

Done

nschloe commented on test/test_cleanfigure.py:
```diff
@@ -0,0 +1,1049 @@
+        plt.close("all")
+
+    def test_plot3d(self):
+        """ """
```
You can remove all those empty doc strings. (And perhaps also the octave-code ones, unless you use them yourself.)

nschloe commented:
> In general one should not test implementation details. I don't want to be too strict about this now, let's keep it in if you want.

I thought about it some more and changed my mind. I want to be strict. :smirk_cat: 

What happens when you replace `tikzplotlib.cleanfigure._something` by plain `cleanfigure` calls as any user would do? Doesn't it take the code path you intend?

Aikhjarto commented:
I just tried this awesome contribution! Thank you for working on it.

`clean_figure()` works nice for a low number (N) of points, but there seems to be an intermediate step that requires large amounts (N^2) of memory.
```python
import matplotlib.pyplot as plt
import numpy as np
plt.plot(np.arange(100000))
tikzplotlib.clean_figure()
```
results in trying to allocate 75 GiB of memory
```

  File "<ipython-input-61-9a9ad396cc52>", line 2, in <module>
    tikzplotlib.clean_figure()

  File "tikzplotlib\tikzplotlib\_cleanfigure.py", line 99, in clean_figure
    fig, target_resolution=target_resolution, scale_precision=scale_precision

  File "tikzplotlib\tikzplotlib\_cleanfigure.py", line 126, in _recursive_cleanfigure
    scale_precision=scale_precision,

  File "tikzplotlib\tikzplotlib\_cleanfigure.py", line 143, in _recursive_cleanfigure
    scale_precision=scale_precision,

  File "tikzplotlib\tikzplotlib\_cleanfigure.py", line 219, in _cleanline
    _prune_outside_box(fighandle, axhandle, linehandle)

  File "tikzplotlib\tikzplotlib\_cleanfigure.py", line 538, in _prune_outside_box
    segvis = _segment_visible(data, dataIsInBox, xLim, yLim)

  File "tikzplotlib\tikzplotlib\_cleanfigure.py", line 952, in _segment_visible
    mask = np.zeros((n - 1, n)) == 1

MemoryError: Unable to allocate 74.5 GiB for an array with shape (99999, 100000) and data type float64
```
There might be room for optimization.

nschloe commented:
Nice catch. @theRealSuperMario Do you have an idea of how to fix this?

theRealSuperMario commented:
mask is a boolean array, but it is created in two steps by allocating an
array of one's and then checking equality with 0.

I think it should be possible to just create the boolean array directly,
thus changing the data type from float64 (64bits) to Boolean (1bit). This
should reduce the memory by a factor of 64.

If it doesn't work like that I will have to think harder.


Besides, I won't be able to fix this until next week. So please give me
some time, also for the tests.



On Fri, Jan 24, 2020, 11:09 Nico Schlömer <notifications@github.com> wrote:

> Nice catch. @theRealSuperMario <https://github.com/theRealSuperMario> Do
> you have an idea of how to fix this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nschloe/tikzplotlib/pull/382?email_source=notifications&email_token=ABRXFWJZHIWWYQEDQHBUJJ3Q7K433A5CNFSM4KI3TEC2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJ2J2GY#issuecomment-578067739>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABRXFWKIZGFTZRULSDLR7F3Q7K433ANCNFSM4KI3TECQ>
> .
>

Aikhjarto commented:
> mask is a boolean array,

Of which size? You allocate it with shape (N-1, N) as default but for `n>1` at the end of `_segment_visible()`  mask is just of shape (N,).
So maybe just the allocation in the default case (n==0) is too large?

nschloe commented:
I agree, a 64-time reduction isn't going to cut it. The issue is that an n-by-n matrix is allocated, independently of the data type.

theRealSuperMario commented:
> Of what size? You allocate it with shape (N-1, N) as default but for `n>1` at the end of `_segment_visible()` mask is just of shape (N,).
> So maybe just the allocation in the default case (n==0) is too large?

Turns out this is just a bug I introduced. The issue is fixed now (with an appropriate test).

theRealSuperMario commented:
> I thought about it some more and changed my mind. I want to be strict. 
> 
> What happens when you replace `tikzplotlib.cleanfigure._something` by plain `cleanfigure` calls as any user would do? Doesn't it take the code path you intend?

Hmm.
I would prefer to test the output and input of every subroutine individually. It was actually very helpful during writing the code to have tests with this level of granularity.

Take for example `_prune_outside_box()`. The method can be summarized by the following pseudocode:

```python
def _prune_outside_box():
    dataIsInBox = _isInBox()
    segvis = _segmentVisible()
    _replace_data_with_NaN()
   _remove_data()
    _remove_NaNs()
```
If I only test the final output, I cannot be sure where the problem occured. If I test every method, it's clear where the problem is.

My suggestion would be to simply [mark the tests](https://stackoverflow.com/questions/52246154/python-using-pytest-to-skip-test-unless-specified) and not run them unless you want to.

```python
py.test --all # will also run tests on protected members
```

In the future, I would adopt the test cases from the matlab2tikz repository. This should provide enough diversity to check all possible paths through the code. Then, I agree I would remove the tests for protected members. Currently, all tests check the same example plot.

That said, maybe it's just too early to merge the PR into master. I just wanted to get feedback early.

theRealSuperMario commented:
> Hmm.
> I would prefer to test the output and input of every subroutine individually. It was actually very helpful during writing the code to have tests with this level of granularity.

Well, I implemented the scatter plot support and realized that I had to change most interfaces, which made most tests obsolte. Therefore, just forget what I wrote above. 

I simplified the testing. I decided to keep one test for one protected method which is `_prune_outside_box`. This call goes through all important subroutines which makes it good enough to spot issues. I still believe it would be a good idea to mark this test as excluded by default and only run it if desired.

In terms of functionality, I am not intending to add any more features because line plots and scatter plots probably already cover 90% of use-cases.

nschloe commented:
Can we get the import to read
```
from tikzplotlib import get_tikz_code, clean_figure
```
? This is how people should use it.

> I simplified the testing. 

Cool.

> I decided to keep one test for one protected method which is _prune_outside_box.

How do you intend the prune feature to be used? A separate user call?

theRealSuperMario commented:
> Can we get the import to read
> 
> ```
> from tikzplotlib import get_tikz_code, clean_figure
> ```

I think I already added this here
https://github.com/nschloe/tikzplotlib/pull/382/files#diff-e1fe340b40b1eb8a9bf4a2e8b9777a7aR23

> How do you intend the prune feature to be used? A separate user call?

Not at all. This just for me to keep the underlying functionality tested. This is kind of a compromise between
> thought about it some more and changed my mind. I want to be strict. smirk_cat
and 
> I would prefer to test the output and input of every subroutine individually. It was actually very helpful during writing the code to have tests with this level of granularity.
by reducing it just to one single test.

The user just uses `clean_figure`, which does pruning and everything else automatically.

nschloe commented:
> I think I already added this here

Then the tests should import and use `clean_figure` just the same way. The tests are a "documentation" of how to use tikzplotlib.

> This is kind of a compromise

There's no room for compromise here. If you want to test `_prune`, use it via `clean_figure`. It should call `_prune`, right? If you cannot be sure that it gets called, something is wrong.

theRealSuperMario commented:
> Then the tests should import and use `clean_figure` just the same way. The tests are a "documentation" of how to use tikzplotlib.

I see. Done.

> There's no room for compromise here

I see that, too. Also done.

nschloe commented:
Okay, this is looking really good now. Got to say one of the finest PRs to tikzplotlib ever.

@theRealSuperMario  If you think it's good to go, I'll merge.

theRealSuperMario commented:
> Okay, this is looking really good now. Got to say one of the finest PRs to tikzplotlib ever.

Well, it was a "schwere Geburt", as we say in German. Thanks for being patient with me until I got it through.

> @theRealSuperMario If you think it's good to go, I'll merge.

Please go ahead.

nschloe commented:
> Well, it was a "schwere Geburt", as we say in German.

Try contributing to numpy or scipy. :) Thanks for pulling through with this!

jebej commented:
Great work! As far as I tried, this does not work with loglog plots, would be awesome if it could!

theRealSuperMario commented:
> Great work! As far as I tried, this does not work with loglog plots, would be awesome if it could!

See for log plots #415
